x-pack/metricbeat/module/gcp/vertexai_logs/vertexai_logs.go

```diff
@@ -0,0 +1,261 @@
+		opt = []option.ClientOption{option.WithCredentialsFile(m.config.CredentialsFilePath)}
+	} else if m.config.CredentialsJSON != "" {
+		opt = []option.ClientOption{option.WithCredentialsJSON([]byte(m.config.CredentialsJSON))}
+	} else {
+		return errors.New("no credentials_file_path or credentials_json specified")
+	}
+
+	client, err := bigquery.NewClient(ctx, m.config.ProjectID, opt...)
+	if err != nil {
+		return fmt.Errorf("error creating bigquery client: %w", err)
+	}
+	defer client.Close()
+
+	datasetID := getDatasetID(m.config.TableID)
+	dataset := client.Dataset(datasetID)
+	meta, err := dataset.Metadata(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting dataset metadata: %w", err)
+	}
+
+	events, err := m.queryVertexAILogs(ctx, client, meta.Location)
+	if err != nil {
+		return fmt.Errorf("queryVertexAILogs failed: %w", err)
+	}
+
+	m.logger.Debugf("Total %d events created for vertexai_logs", len(events))
+	for _, event := range events {
+		reporter.Event(event)
+	}
+
+	// Update watermark with latest logging_time from events
+	m.updateLastLoggingTime(events)
+
+	return nil
+}
+
+func (m *MetricSet) queryVertexAILogs(ctx context.Context, client *bigquery.Client, location string) ([]mb.Event, error) {
+	query := m.generateQuery()
+	m.logger.Debug("bigquery query = ", query)
+
+	q := client.Query(query)
+	q.Location = location
+
+	job, err := q.Run(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("bigquery Run failed: %w", err)
+	}
+
+	status, err := job.Wait(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("bigquery Wait failed: %w", err)
+	}
+
+	if err := status.Err(); err != nil {
+		return nil, fmt.Errorf("bigquery status error: %w", err)
+	}
+
+	it, err := job.Read(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("reading from bigquery job failed: %w", err)
+	}
+
+	var rows []VertexAILogRow
+	for {
+		var row VertexAILogRow
+		err := it.Next(&row)
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("bigquery RowIterator Next failed: %w", err)
+		}
+		rows = append(rows, row)
+	}
+
+	events := EventsMapping(rows, m.config.ProjectID, m.logger)
+	return events, nil
+}
+
+func (m *MetricSet) generateQuery() string {
+	escapedTableID := fmt.Sprintf("`%s`", m.config.TableID)
+
+	var whereClause string
+	if m.lastLoggingTime != nil {
+		// Incremental query: get records after last processed time
+		whereClause = fmt.Sprintf("logging_time > TIMESTAMP('%s')",
+			m.lastLoggingTime.Format("2006-01-02 15:04:05.000000"))
+		m.logger.Debugf("Using incremental query from logging_time: %s", m.lastLoggingTime.Format(time.RFC3339))
+	} else {
+		// First run: use timestamp filter with lookback
+		whereClause = fmt.Sprintf("logging_time >= TIMESTAMP_SUB(CURRENT_TIMESTAMP(), INTERVAL %d HOUR)",
+			m.config.TimeLookbackHours)
+		m.logger.Debugf("Using initial query with timestamp filter: %d hours", m.config.TimeLookbackHours)
+	}
+
+	query := fmt.Sprintf(`
+SELECT
+	IFNULL(endpoint, '') AS endpoint,
```

Comment:
do we need to use `IFNULL` here vs using the built in nullable types in the client lib e.g. `bigquery.NullString`

Reply:
i think the `Null*` types handle their own unmarshaling though, so i doubt null checks would be required in code? is there any perf implications? (in general pushing this stuff to BQ is the right move, but just want to make sure it doesn't significantly slow down queries)

Reply:
I am not sure wrt performance but we will have to access and validate it like below. 
If you see [here](http://github.com/googleapis/google-cloud-go/blob/main/bigquery/nulls.go#L48) . 
We will have to check for valid and then get the value with StringVal.
https://cloud.google.com/go/docs/reference/cloud.google.com/go/bigquery/latest#cloud_google_com_go_bigquery_NullString .

Do you think we should go with BigQuery NullString ?